examples/api-provider-sample/src/node/gotd-backend-module.ts

```diff
@@ -0,0 +1,25 @@
+// *****************************************************************************
+// Copyright (C) 2023 EclipseSource and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// http://www.eclipse.org/legal/epl-2.0.
+//
+// This Source Code may also be made available under the following Secondary
+// Licenses when the conditions for such availability set forth in the Eclipse
+// Public License v. 2.0 are satisfied: GNU General Public License, version 2
+// with the GNU Classpath Exception which is available at
+// https://www.gnu.org/software/classpath/license.html.
+//
+// SPDX-License-Identifier: EPL-2.0 OR GPL-2.0-only WITH Classpath-exception-2.0
+// *****************************************************************************
+import { ContainerModule } from '@theia/core/shared/inversify';
+import { ExtPluginApiProvider } from '@theia/plugin-ext';
+import { ExtPluginGotdApiProvider } from './ext-plugin-gotd-api-provider';
+import { MainPluginApiProvider } from '@theia/plugin-ext/lib/common/plugin-ext-api-contribution';
+import { GotdMainPluginApiProvider } from './gotd-main-plugin-provider';
+
+export default new ContainerModule(bind => {
+    bind(Symbol.for(ExtPluginApiProvider)).to(ExtPluginGotdApiProvider).inSingletonScope();
```

Comment:
How do we specify here that this provider should be providing an API object only in the headless plugin-host and not the traditional plugin-host? In other words, if this `api-provider-sample` Theia Extension wanted to contribute namespace ABC to traditional plugins (VS Code Extensions) and namespace DEF to headless plugins, is that not possible?

Reply:
Disregard this comment. As I dove further into having a **single** Theia Extension provide different API namespaces for traditional plugins vs headless one, I see that it's not really feasible. So, I think the moral of the story here is: in that situation, the Theia application developer should have two Theia Extensions--one that provides the traditional API, one that provides the headless API.

Reply:
I’ve experimented with this PR and I believe I've confirmed my suspicions. There is no scoping of API providers to a specific plugin-host (normal vs headless). A contributed API provider is instantiated and exercised in both types of plugin-hosts. This seems problematic to me. It means headless plugin can do an import of a normal (non-headless) custom API module and the corresponding API provider running in the headless plugin-host will give it to him--even if it makes no sense. The headless plugin will end up with an API object whose calls will never complete because the Ext object has a proxy to something that has no Main on the other end. 


But “wait”, you say. “Why is a headless plugin trying to interact with the UI? That’s not what headless plugins are for.” Absolutely. The current implementation in this PR is not responsible for that nonsense. It is, however, responsible for allowing the nonsense to get too far. The import of the normal API object from a headless plugin should fail but it doesn’t.


There is a case where this actually would be desirable, though. Technically speaking, an API object need not interact with anything on the mainland. An app developer could theoretically provide an API object whose implementation can be carried out entirely in the plugin-host–i.e., one that does not make use of RPC. E.g., a Math API object, as silly as that would be. In that case, there would be an argument for an API provider to allow itself to be available for both types of plugins: normal and headless one. Should we support that? If so…sure; that’s fine. But we still need the scoping to prevent it in the 99% of situations, where it’s not going to make sense.

Reply:
The distinction of which plugin host an API provider may publish its API should now be possible with commit f100932d6c68cd309a3ba08adcd82531012bb72c.